controllers/jenkins/pipelinerun/pipelinerun_controller.go

```diff
@@ -39,12 +49,175 @@ type Reconciler struct {
+		return ctrl.Result{}, nil
+	}
+
+	var pipeline v1alpha3.Pipeline
+	if err := r.Client.Get(ctx, client.ObjectKey{Namespace: pr.GetNamespace(), Name: pipelineRef.Name}, &pipeline); err != nil {
+		log.Error(err, "unable to get pipeline")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	var projectName = pipeline.Namespace
+
+	// check pipeline run status
+	if pr.HasStarted() {
+		log.V(5).Info("pipeline has already started, and we are retrieving pipeline run data from Jenkins.")
+
+		runResult, err := r.getPipelineRunResult(projectName, pipeline.GetName(), &pr)
+		if err != nil {
+			log.Error(err, "unable get pipeline run data.")
+			return ctrl.Result{}, err
+		}
+
+		// set latest run result into annotations
+		if pr.Annotations == nil {
+			pr.Annotations = make(map[string]string)
+		}
+		runResultJSON, err := json.Marshal(runResult)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		pr.Annotations[devopsv1alpha4.JenkinsPipelineRunDataKey] = string(runResultJSON)
+
+		// update pipeline run
+		if err := r.Update(ctx, &pr); err != nil {
+			log.Error(err, "unable to update pipeline run.")
+			return ctrl.Result{}, err
+		}
 
-	// TODO Remaining logic
+		// TODO Update pipeline run status according to run result
+		if runResult.EndTime != "" {
+			// this means the pipeline run has ended
+			return ctrl.Result{}, nil
+		}
+
+		// until the status is okay
+		// TODO make the RequeueAfter configurable
+		return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+	}
+
+	// first run
+	parameters := pr.Spec.Parameters
+	if parameters == nil {
+		parameters = make([]devopsv1alpha4.Parameter, 0)
+	}
+
+	// build http parameters
+	httpParameters, err := buildHttpParametersForRunning(&pr)
+	if err != nil {
+		log.Error(err, "unable to create http parameters for running pipeline.")
+		return ctrl.Result{}, err
+	}
+	// run pipeline
+	var runResponse *devopsClient.RunPipeline
+	if pr.IsMultiBranchPipeline() {
+		// run multi branch pipeline
+		runResponse, err = r.DevOpsClient.RunBranchPipeline(projectName, pipeline.GetName(), pr.Spec.SCM.RefName, httpParameters)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	} else {
+		// run normal pipeline
+		runResponse, err = r.DevOpsClient.RunPipeline(projectName, pipeline.GetName(), httpParameters)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// label pipelinerun as running
+	if pr.Annotations == nil {
+		pr.Annotations = make(map[string]string)
+	}
+	pr.Annotations[devopsv1alpha4.JenkinsPipelineRunIdKey] = runResponse.ID
+	pr.Status.MarkPending()
+	now := v1.NewTime(time.Now())
+	pr.Status.StartTime = &now
+	pr.Status.UpdateTime = &now
+
+	if err := r.Client.Update(ctx, &pr); err != nil {
+		log.Error(err, "unable to update pipeline run")
+		return ctrl.Result{}, err
+	}
+	if err := r.Client.Status().Update(ctx, &pr); err != nil {
+		log.Error(err, "unable to update pipeline run status.")
+		return ctrl.Result{}, err
+	}
```

Comment:
Why update it twice?

Reply:
I've add reason in code comment as below:

```
// Because the status is a subresource of PipelineRun, we have to update status separately.
// See also: https://book-v1.book.kubebuilder.io/basics/status_subresource.html
```